### ISSUE TITLE
Better Event Hooks

### ISSUE THREAD

**Issue by w-** [foxhound87/mobx-ajv-form#172]
Hi,
Starting to use mobx and have a requirement for forms. mobx-react-form looks very comprehensive. Thank you for your hard work on it.

My use case is that I have a form and want to use the data in the form to drive a a real-time preview. i.e. user inputs data in form and can see in realtime  how their inputs affect the preview.

Previously I made a simple store to accomplish this. In this model, both the react component for the form UI and the preview accepted the same object as a prop, i.e. my simple store.

If I use a mobx-react-form, how do I drive the preview?
The preview is a reusable component that expects as prop an object with specific structure. 
so for example I'd be able to do this:
```
// simpleStore has observable attribute "title"
<PreviewComponent blah={simpleStore}/>
// or
<PreviewComponent blah={otherObjectWithSameStructure}/>
//
//....
//
// then inside previewComponent might be something like
return <div>{props.blah.title}</div>;
```

in the context of mobx-react-form, how would i accomplish the same? i.e. I want to pass the form to PreviewComponent but i want the fields to be observable.

this seems super basic, I know I must be missing something silly.

**Comment by foxhound87** [comment]
Hi @w- 

this package already is a form store.

What you want to achieve should be easy to do, 

You need to pass the form instance down two of your children components,
for example `<MyForm form={form} />` and `<MyPreview form={form} />`

then use for example `form.$('password').value` to select your fields and render the values. 
(the `value` property is a mobx computed value). 

just make sure to [decorate the components with the mobx observer](https://github.com/foxhound87/mobx-react-form-demo/blob/master/src/components/inputs/SimpleInput.jsx#L4)

See the [Available Field Properties](https://foxhound87.github.io/mobx-react-form/docs/api-reference/fields-properties.html) on the docs.

**Comment by foxhound87** [comment]
Did my suggestions helped?

**Comment by foxhound87** [comment]
I think I can close this for inactivity. 
Feel free to reopen if you encounter other issues.

**Comment by w-** [comment]
@foxhound87 
Hi, thank you for your response and sorry for my late reply. 

the challenge with your suggestion is that my the MyPreview component then must accept an observable object that looks like a form.

i.e. currently an existing MyPreview component accepts an observable object that looks like
```
@observable thing = {
	title: 'blah',
	detail: 'foobar'
}

```
and it will render something say with

`{{ thing.title }}`

so i cannot pass a form that has the fields title and detail because accessing the values for these fields would only be available via doing what you have in your example

`{{ form.$('title').value }}`

Does this help clarify what i'm getting at?

**Comment by foxhound87** [comment]
then you should able to pass directly the object to your component like:

```javascript
<MyPreview myProp={{
  title: form.$('title').value,
  detail: form.$('detail').value,
}} />
```

**Comment by w-** [comment]
yes i understand that. ideally i'm trying to avoid repeating myself by recreating the structure but from your replies i guess it wont be possible. 

is there an event on a form that is fired when any field's value changes?
I know there are events per field, but i'm looking for a form-wide one.
the idea would be that i could just listen for that event and do a values() call on the form whenever something changes. (not exactly the most efficient i know)

**Comment by foxhound87** [comment]
I don't know how your data and components are structured in detail, that's just an example.

If you build your components around the mobx-react-form apis you will reduce the amount of the boilerplate, as the package has its own data structure, but if you want to use components which have a pre-defined structure, you definitively have to rethink the data structure/management too,
but I think the effort can be less painful using this package's apis instead to implement all from scratch.

The events hooks introduced in the current version are a limited implementation and don't work well for all situation and don't have test cases implemented yet. 

In the future I will release a better event handling/hooks methods.
I reopen this issue as `enhancement` scheduled for the next major release.

**Comment by foxhound87** [comment]
Anyway, in the meantime you can use the mobx `observe` to listen on the prop `$value` changes.

Something like this:

```javascript
import { observe } from 'mobx';

class MyForm extends MobxReactForm {

  onInit(form) {
    
    form.fields.forEach((field, key) => {
      observe(form.fields.get(key), '$value', () => {
        console.log(field.value);
      });
    });
  }
}
```

Using the built-in package's `forEach()` you can iterate nested fields too, 
and the implementation is even more compact:

```javascript
import { observe } from 'mobx';

class MyForm extends MobxReactForm {

  onInit(form) {
    
    // this will iterate nested fields automatically
    form.forEach((field) => {
      observe(field, '$value', () => {
        console.log(field.value);
      });
    });
  }
}
```

**Comment by anotherpit** [comment]
@foxhound87, doesn't `form.values()` return the structure @w- needs?

```jsx
<MyPreview myProp={{form.values()}} />
```

**Comment by foxhound87** [comment]
@anotherpit Yes, but I think he wants only get some of them to pass to a component with a specific structure. @w- is it right?

**Comment by foxhound87** [comment]
@anotherpit anyway, In that way the values are not observable, and the component will not re-render on changes.

**Comment by w-** [comment]
@foxhound87 
Thank you. Thats an interesting suggestion you make looping through each field and making each value observable in that manner. I'll look further into that.

@anotherpit 
> @anotherpit anyway, In that way the values are not observable, and the component will not re-render on changes.

yes, the specific issue is that the returned values are not observable. this is why i  cannot' use form.values()

I'm happy that form.values() doesn't return observables, its just that it isn't appropriate for my specific use case.

**Comment by foxhound87** [comment]
Implemented in [version 1.19](https://github.com/foxhound87/mobx-react-form/commit/005d04c7)

[Events Hooks](https://foxhound87.github.io/mobx-react-form/docs/events/events-hooks.html)
[Observe Fields Props](https://foxhound87.github.io/mobx-react-form/docs/events/observe.html)
